mentat/code_context.py

```diff
@@ -147,188 +60,313 @@ class CodeContextSettings:
+
+    async def _get_auto_features(
+        self,
+        model: str,
+        include_features: list[CodeFile],
+        max_tokens: int,
+    ) -> list[CodeFile]:
+        git_root = GIT_ROOT.get()
+        # Generate all possible permutations for all files in the project.
+        candidate_features = list[CodeFile]()
+        for path in get_non_gitignored_files(git_root):
+            if not path.is_dir() and not is_file_text_encoded(path):
+                continue
+            permutations: list[tuple[CodeMessageLevel, str | None]] = [
+                (CodeMessageLevel.CODE, None),
+                (CodeMessageLevel.FILE_NAME, None),
+            ]
+            if not self.settings.no_code_map:
+                permutations += [
+                    (CodeMessageLevel.CMAP_FULL, None),
+                    (CodeMessageLevel.CMAP, None),
+                ]
+            if self.diff_context.target and path in self.diff_context.files:
+                permutations += [
+                    (level, self.diff_context.target) for level, _ in permutations
+                ]
+            for level, diff in permutations:
+                feature = CodeFile(path, level=level, diff=diff)
+                candidate_features.append(feature)
+
+        # Sort candidates by relevance/density.
+        tokens_remaining = max_tokens
+        sem = asyncio.Semaphore(10)
+
+        async def _calculate_feature_score(feature: CodeFile) -> float:
+            score = 0.0
+            async with sem:
+                tokens = await feature.count_tokens(model)
+            if tokens == 0:
+                raise MentatError(f"Feature {feature} has 0 tokens.")
+            if feature.diff is not None:
+                score += 1
+            if feature.level == CodeMessageLevel.FILE_NAME:
+                score += 0.1
+            if feature.level == CodeMessageLevel.CMAP:
+                score += 0.25
+            elif feature.level == CodeMessageLevel.CMAP_FULL:
+                score += 0.5
+            elif feature.level == CodeMessageLevel.CODE:
+                score += 0.75
+            score /= tokens
+            return score
+
+        all_features = include_features.copy()
+        candidate_scores_task = [
+            _calculate_feature_score(f) for f in candidate_features
+        ]
+        candidate_scores = await asyncio.gather(
+            *candidate_scores_task, return_exceptions=True
+        )
+        candidates_scored = list(zip(candidate_features, candidate_scores))
+        candidates_sorted = sorted(candidates_scored, key=lambda x: x[1], reverse=True)
+        for feature, _ in candidates_sorted:
+            feature_tokens = await feature.count_tokens(model)
+            if tokens_remaining - feature_tokens <= 0:
+                continue
+            if _longer_feature_already_included(feature, all_features):
+                continue
+            to_replace = _shorter_features_already_included(feature, all_features)
```

Comment:
This is the only place this is used and this algorithm could include at most one other shorter feature right? Maybe this function could be `_remove_shorter_feature` and do the below modification itself? It would save a for loop.

Reply:
For now that's right, thought in the future I expect a single feature (full-file code) could replace multiple (intervals or individual functions).

Also we need to adjust the running token balance (line 327). I think that'd involve either passing tokens to the helper func (not ideal) or re-counting the entire spliced feature set afterwards, which seems also not ideal.